Convert PlatformEvent timestamps to MonotonicTime

francovs opened the pull request:
#### abb74d61fd74f859586b597d8a6513b29b2af27b
<pre>
Convert PlatformEvent timestamps to MonotonicTime
<a href="https://bugs.webkit.org/show_bug.cgi?id=298552">https://bugs.webkit.org/show_bug.cgi?id=298552</a>
<a href="https://rdar.apple.com/158717958">rdar://158717958</a>

Reviewed by Chris Dumez.

PlatformEvent objects represent user input such as taps, mouse clicks
or key presses, and are used to create used to create DOM UIEvent
objects. The DOM UIEvent class uses MonotonicTime as its time base, so
it is natural for PlatformEvents to do so as well.

This patch avoids an unnecessary round trip: timestamps were being
converted from native timestamps (monotonic) to WallTime, and later to
MonotonicTime. The conversion to WallTime introduces small errors
which cause inconsistencies when comparing performance.now() to event
timestamps.

This patch is not supposed to cause any change in behavior. However,
these changes are not strictly refactors:

* Tools/WebKitTestRunner/mac/EventSenderProxy.mm : issued events now
  contain a timestamp with a small fixed value instead of zero. This is
  required to make sure the timestamp is not interpreted as a falsy
  value in some scrolling logic downstream - an event with a
  zeroed timestamp will produce a MonotonicTime with the falsy, default
  initialized value; this doesn&apos;t happen with WallTime.
* Source/WebCore/platform/ios/PlatformEventFactoryIOS.mm : the
  PlatformTouchEventBuilder(WebEvent *event) constructor used to create
  a WallTime object directly from an event timestamp, with
  WallTime::fromRawSeconds(event.timestamp); the change to
  MonotonicTime::fromRawSeconds(event.timestamp) changes how
  event.timestamp is interpreted (from seconds since 1970 to seconds
  since system boot). The previous behavior was possibly an error, and
  the only code path that uses the affected constructor is the touch()
  method in Source/WebKitLegacy/mac/WebView/WebHTMLView.mm.
* Source/WebCore/platform/win/{Key,Wheel}EventWin.cpp : WallTime event
  timestamps were being created from GetTickCount(), which counts
  milliseconds from boot, creating timestamps close to the WallTime
  epoch (1970?). This was replaced by MonotonicTime::now()

Canonical link: <a href="https://commits.webkit.org/299753@main">https://commits.webkit.org/299753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d90cd59c9a1cfdc99c9404dabe572289ce03274e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119970 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30314 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126300 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72043 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121846 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48240 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91104 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60418 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122922 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32235 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107569 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71660 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31266 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25674 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69935 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101710 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25862 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129220 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46890 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35553 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99729 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47256 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103757 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99574 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25304 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45022 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23033 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43511 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46752 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52458 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46218 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49567 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47904 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
```diff
@@ -142,6 +142,8 @@ - (id)initMagnifyEventAtLocation:(NSPoint)location globalLocation:(NSPoint)globa
     CGEventSetIntegerValueField(cgEvent.get(), kCGEventGesturePhase, EventSenderCGGesturePhaseFromNSEventPhase(phase));
     CGEventSetDoubleValueField(cgEvent.get(), kCGEventGestureZoomValue, magnification);
+    // Set a value that won't be interpreted as a falsy timestamp:
+    CGEventSetTimestamp(cgEvent.get(), 1);
```
Shouldn't we use the provided `time` somehow? Setting to `1` seems unnecessarily "hackish".

francovs replied:
Definitely hackish, but adding a timestamp that updates on every call caused a autoscroll test to fail (`rubberband-axis-locking.html`). Since a fixed timestamp is the current status quo, I want to leave a fix for that to a [follow up PR](https://github.com/WebKit/WebKit/pull/50104).

Also, I'm still not sure if the test breakage that happens when adding the timestamp is a bug or expected behavior... the failing test involve "continuation" scroll events, maybe those are expected to have the same timestamp?

cdumez reviewed (approved):
LGTM

francovs committed:
Convert PlatformEvent timestamps to MonotonicTime
https://bugs.webkit.org/show_bug.cgi?id=298552
rdar://158717958

Reviewed by Chris Dumez.

PlatformEvent objects represent user input such as taps, mouse clicks
or key presses, and are used to create used to create DOM UIEvent
objects. The DOM UIEvent class uses MonotonicTime as its time base, so
it is natural for PlatformEvents to do so as well.

This patch avoids an unnecessary round trip: timestamps were being
converted from native timestamps (monotonic) to WallTime, and later to
MonotonicTime. The conversion to WallTime introduces small errors
which cause inconsistencies when comparing performance.now() to event
timestamps.

This patch is not supposed to cause any change in behavior. However,
these changes are not strictly refactors:

* Tools/WebKitTestRunner/mac/EventSenderProxy.mm : issued events now
  contain a timestamp with a small fixed value instead of zero. This is
  required to make sure the timestamp is not interpreted as a falsy
  value in some scrolling logic downstream - an event with a
  zeroed timestamp will produce a MonotonicTime with the falsy, default
  initialized value; this doesn't happen with WallTime.
* Source/WebCore/platform/ios/PlatformEventFactoryIOS.mm : the
  PlatformTouchEventBuilder(WebEvent *event) constructor used to create
  a WallTime object directly from an event timestamp, with
  WallTime::fromRawSeconds(event.timestamp); the change to
  MonotonicTime::fromRawSeconds(event.timestamp) changes how
  event.timestamp is interpreted (from seconds since 1970 to seconds
  since system boot). The previous behavior was possibly an error, and
  the only code path that uses the affected constructor is the touch()
  method in Source/WebKitLegacy/mac/WebView/WebHTMLView.mm.
* Source/WebCore/platform/win/{Key,Wheel}EventWin.cpp : WallTime event
  timestamps were being created from GetTickCount(), which counts
  milliseconds from boot, creating timestamps close to the WallTime
  epoch (1970?). This was replaced by MonotonicTime::now()

Canonical link: https://commits.webkit.org/299753@main
